secure/secure.sh

```diff
@@ -0,0 +1,403 @@
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+# This script installs and configures Kerberos on a Google Cloud
+# Dataproc cluster.
+
+# Do not use -x to avoid printing passwords.
+set -euo pipefail
+
+# Fail under HA mode early
+additional_nodes=$(/usr/share/google/get_metadata_value attributes/dataproc-master-additional | sed 's/,/\n/g' | wc -l)
+if [[ $additional_nodes -gt 0 ]]; then
+  echo 'Kerberos is not supported on HA mode clusters.'
+  exit 1
+fi
+
+function update_apt_get() {
+  for ((i = 0; i < 10; i++)); do
+    if apt-get update; then
+      return 0
+    fi
+    sleep 5
+  done
+  return 1
+}
+
+function err() {
+  echo "[$(date +'%Y-%m-%dT%H:%M:%S%z')]: $@" >&2
+  return 1
+}
+
+ROLE="$(/usr/share/google/get_metadata_value attributes/dataproc-role)"
+WORKER_COUNT="$(/usr/share/google/get_metadata_value attributes/dataproc-worker-count)"
+FQDN="$(hostname -f | tr -d '\n')"
+DOMAIN=${FQDN#*\.}
+REALM=$(echo $DOMAIN | awk '{print toupper($0)}')
+MASTER="$(/usr/share/google/get_metadata_value attributes/dataproc-master)"
+MASTER_FQDN="$MASTER.$DOMAIN"
+HADOOP_CONF_DIR="/etc/hadoop/conf"
+
+readonly kms_key_uri=$(/usr/share/google/get_metadata_value attributes/kms-key-uri)
+readonly db_password_uri=$(/usr/share/google/get_metadata_value attributes/db-password-uri)
+readonly root_password_uri=$(/usr/share/google/get_metadata_value attributes/root-password-uri)
+readonly keystore_password_uri=$(/usr/share/google/get_metadata_value attributes/keystore-password-uri)
+readonly db_password=$(gsutil cat $db_password_uri | \
+  gcloud kms decrypt \
+  --ciphertext-file - \
+  --plaintext-file - \
+  --key $kms_key_uri)
+readonly root_principal_password=$(gsutil cat $root_password_uri | \
+  gcloud kms decrypt \
+  --ciphertext-file - \
+  --plaintext-file - \
+  --key $kms_key_uri)
+readonly keystore_password=$(gsutil cat $keystore_password_uri | \
+  gcloud kms decrypt \
+  --ciphertext-file - \
+  --plaintext-file - \
+  --key $kms_key_uri)
+readonly dataproc_bucket=$(/usr/share/google/get_metadata_value attributes/dataproc-bucket)
+readonly keystore_uri=$(/usr/share/google/get_metadata_value attributes/keystore-uri)
+readonly truststore_uri=$(/usr/share/google/get_metadata_value attributes/truststore-uri)
+readonly cluster_uuid=$(/usr/share/google/get_metadata_value attributes/dataproc-cluster-uuid)
+readonly krb5_server_mark_file="krb5-server-mark"
+
+function install_and_start_kerberos_server() {
+  DEBIAN_FRONTEND=noninteractive apt-get install -y krb5-kdc; \
+  DEBIAN_FRONTEND=noninteractive apt-get install -y krb5-admin-server || TRUE
```

Comment:
Why is it okay if this fails?

Reply:
changed. If installation fails scrip will fail.